index.js

```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
   })
 }
 
+function getSerializedKeyTokens (key) {
+  const tokens = key.split(/(\[|]|{})/).filter(Boolean)
+  return tokens.some(token => ['[', ']', '{}'].includes(token)) ? tokens : []
+}
+
+function validateSerializedKey (tokens, value) {
+  let isValid = true
+
+  // first token can't be '[', ']', or '{}'
+  if (['[', ']', '{}'].includes(tokens[0])) isValid = false
+  // only a single {} special ending is allowed
+  else if (tokens.filter(t => t === '{}').length > 1) isValid = false
+  // {}, if present, must be the last token
+  else if (tokens.find(t => t === '{}') && tokens[tokens.length - 1] !== '{}') isValid = false
+
+  let hasOpenBracket = false
+  for (let i = 1; i < tokens.length; i++) {
+    const token = tokens[i]
```

Comment:
maybe define token and nextToken as let outside the for loop. Avoiding variable instantiation by reusing the variables.


Someting like:
```js
const tokensLength = tokens.length
let token = tokens[0]
let nextToken = tokens[1]
let i = 1
 while (i < tokensLength) {
     token = nextToken
     nextToken = tokes[++i]
```

Reply:
Even though, IMO, this approach seems uglier, I have to agree that [it performs better](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJSZWd1bGFyIGZvciBsb29wICsgYmxvY2sgbGV2ZWwgY29uc3RzIFZTIHdoaWxlIGxvb3AgKyBtYW51YWxseSBob2lzdGluZyB2YXJzIiwiYmVmb3JlIjoiY29uc3QgZGF0YSA9IG5ldyBBcnJheSgxMDAwKS5maWxsKDApLm1hcCgoKSA9PiBNYXRoLnJhbmRvbSgpLnRvU3RyaW5nKCkpOyIsInRlc3RzIjpbeyJuYW1lIjoiUmVndWxhciBmb3IgbG9vcCArIGJsb2NrIGxldmVsIGNvbnN0cyIsImNvZGUiOiJmb3IgKGxldCBpID0gMDsgaSA8IGRhdGEubGVuZ3RoOyBpKyspIHtcbiAgY29uc3QgdmFsdWUgPSBkYXRhWzBdO1xuICBjb25zdCBuZXh0VmFsdWUgPSBkYXRhWzFdO1xufSIsInJ1bnMiOls2MDAwLDIwMDAsNjAwMCwzMDAwLDIwMDAsNTAwMCwxODAwMCwxMjAwMCwxMzAwMCw1MDAwLDYwMDAsMTEwMDAsMTcwMDAsNTAwMCw2MDAwLDIyMDAwLDIzMDAwLDE1MDAwLDYwMDAsMTAwMCw3MDAwLDYwMDAsMTAwMDAsMjQwMDAsMTcwMDAsMjAwMCw3MDAwLDIwMDAwLDEwMDAsMTQwMDAsNTAwMCw5MDAwLDYwMDAsNTAwMCwxNzAwMCwxMjAwMCwxODAwMCwxNjAwMCwxMzAwMCw3MDAwLDUwMDAsNTAwMCwxMzAwMCw0MDAwLDYwMDAsMTkwMDAsMzAwMCwxMjAwMCwxODAwMCwxMTAwMCw3MDAwLDIzMDAwLDQwMDAsMTAwMCw5MDAwLDEyMDAwLDE2MDAwLDIzMDAwLDE3MDAwLDEwMDAsMTAwMCwxNjAwMCw2MDAwLDIwMDAwLDEwMDAwLDkwMDAsMTMwMDAsMTYwMDAsMTAwMDAsMTAwMDAsNjAwMCw2MDAwLDE0MDAwLDQwMDAsNzAwMCw3MDAwLDkwMDAsNTAwMCw2MDAwLDE2MDAwLDQwMDAsNjAwMCw0MDAwLDIwMDAsMTAwMDAsOTAwMCw4MDAwLDUwMDAsNjAwMCwyMjAwMCwxMDAwLDExMDAwLDI0MDAwLDIxMDAwLDE3MDAwLDIwMDAsNDAwMCw1MDAwLDIxMDAwLDQwMDBdLCJvcHMiOjk4NjB9LHsibmFtZSI6IndoaWxlIGxvb3AgKyBtYW51YWxseSBob2lzdGluZyB2YXJzIiwiY29kZSI6ImNvbnN0IGRhdGFMZW5ndGggPSBkYXRhLmxlbmd0aDtcbmxldCB2YWx1ZSA9IGRhdGFbMF07XG5sZXQgbmV4dFZhbHVlID0gZGF0YVsxXTtcbmxldCBpID0gMTtcbndoaWxlIChpIDwgZGF0YUxlbmd0aCkge1xuICAgdmFsdWUgPSBkYXRhW2ldO1xuICAgbmV4dFZhbHVlID0gZGF0YVsrK2ldO1xufSIsInJ1bnMiOls0MDAwLDkwMDAsODAwMCw3MDAwLDkwMDAsNDAwMCw5MDAwLDIxMDAwLDQwMDAsMTIwMDAsMTkwMDAsMTAwMDAsOTAwMCwyMTAwMCw4MDAwLDgwMDAsNTAwMCwxODAwMCwyNjAwMCwyNjAwMCwxNzAwMCwxMDAwMCwxNTAwMCwyMzAwMCw0MDAwLDE3MDAwLDIwMDAsMjAwMDAsMTMwMDAsMTcwMDAsOTAwMCwyMTAwMCwyMDAwLDI1MDAwLDEzMDAwLDcwMDAsMjAwMDAsOTAwMCwxNjAwMCw5MDAwLDQwMDAsNjAwMCwyNDAwMCw3MDAwLDI1MDAwLDEwMDAwLDgwMDAsMTUwMDAsNDAwMCw5MDAwLDYwMDAsMTAwMCwxNTAwMCwxMzAwMCwxMjAwMCw5MDAwLDMwMDAsMjUwMDAsMTgwMDAsODAwMCwxMDAwLDYwMDAsOTAwMCwyMDAwLDEwMDAsMTkwMDAsNDAwMCwxNjAwMCwzMDAwLDEyMDAwLDI1MDAwLDkwMDAsMjEwMDAsNDAwMCwyNDAwMCwzMDAwLDE4MDAwLDQwMDAsMjIwMDAsMTQwMDAsMzAwMCw5MDAwLDkwMDAsMTUwMDAsMzAwMCw4MDAwLDkwMDAsMTgwMDAsNjAwMCwyMzAwMCw5MDAwLDYwMDAsMjkwMDAsNzAwMCwxMDAwLDE0MDAwLDEyMDAwLDIxMDAwLDExMDAwLDQwMDBdLCJvcHMiOjExNjIwfV0sInVwZGF0ZWQiOiIyMDIzLTA1LTExVDIyOjMzOjE2LjM2MFoifQ%3D%3D). And since performance is a main concern for Fastify and its plugins, I've changed it.